clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -783,8 +783,10 @@ public void handle(HeartbeatResponse heartbeatResponse, RequestFuture<Void> futu
         @Override
         public void onFailure(RuntimeException e, RequestFuture<T> future) {
             // mark the coordinator as dead
-            if (e instanceof DisconnectException)
+            if (e instanceof DisconnectException) {
                 coordinatorDead();
+                requestRejoin();
```

Comment:
I think @hachikuji 's suggestion may be better: do not call 

```
AbstractCoordinator.this.rejoinNeeded = false;
```

in `JoinGroupResponseHandler#handle()`, but in `SyncGroupResponseHandler#handle()`.